### PR TITLE
Add limited support for non-equi outer joins

### PR DESCRIPTION
**Pull request by losipiuk**
Add support for non-equi conjuncts in join condition for outer joins.
Supported case is when conjunct is resolvable as a whole using symbols
of inner join table. In such case conjunct is pushed down as filter to
inner source relation.

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -1805,18 +1804,34 @@ public void testNonEqualityJoin()
...
+        assertQuery("SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey AND orders.custkey > 1000");
+        assertQuery("SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey AND orders.custkey > orders.shippriority");
+    }
+
     @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = ".*Non-equi.*")
-    public void testNonEqualityLeftJoin()
+    public void testUnsupportedNonEqualityLeftJoin()
+            throws Exception
+    {
+        assertQuery("SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey AND orders.custkey > lineitem.quantity");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: use assertQueryFails

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1187,6 +1187,24 @@ else if (criteria instanceof JoinOn) {
...
+                        continue;
+                    }
+                }
+                if (node.getType() == Join.Type.LEFT) {
+                    Set<QualifiedName> names = DependencyExtractor.extractNames(conjunct, analyzer.getColumnReferences());
+                    if (names.stream().allMatch(right.canResolvePredicate())) {
+                        analyzeExpression(conjunct, right, context);
+                        continue;
+                    }
+                }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1206]: what about FULL OUTER?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -1805,18 +1804,34 @@ public void testNonEqualityJoin()
         assertQuery("SELECT COUNT(*) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey AND NULL");
     }
 
+    @Test
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: add test for FULL OUTER

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 4]
```diff
@@ -239,8 +244,27 @@ protected RelationPlan visitJoin(Join node, Void context)
...
+                // Criteria expression looks like "a = a" which then seem to be resolvable by any side of join.
+                if ((node.getType() == LEFT || node.getType() == RIGHT) && (!(node.getCriteria().get() instanceof JoinUsing))) {
+                    Set<QualifiedName> dependencies = DependencyExtractor.extractNames(conjunct, analysis.getColumnReferences());
+                    if (node.getType() == LEFT) {
+                        if (dependencies.stream().allMatch(right.canResolvePredicate())) {
+                            preRightConditions.add(rightPlanBuilder.rewrite(conjunct));
+                            continue;
+                        }
+                    }
+                    else {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: why not `else if (node.getType == RIGHT)`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 5]
```diff
@@ -1805,18 +1804,34 @@ public void testNonEqualityJoin()
...
+    public void testSupportedNonEqualityLeftJoin()
+            throws Exception
+    {
+        assertQuery("SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey AND orders.custkey > 1000");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: what about coercions? Can you try to test:

```
SELECT COUNT(*) FROM lineitem LEFT OUTER JOIN orders ON lineitem.orderkey = orders.orderkey AND orders.custkey > 1000.0
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1206]: it is not supported. I will add test

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 4 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: this is always the case I will add that anyway

**Comment by losipiuk** [comment]
superseded by https://github.com/facebook/presto/pull/4796

**Pull request closed without merging**